### PR TITLE
WIP: Hide recurrence tags in displayed task string

### LINKED ISSUE #288: Recurrence tags
Possible to (optionally) hide the 'rec:1w' tags when displaying tasks in widget?

### PR DESCRIPTION
**Pull request by TiiXel**
This should close #288, it hides recurrence tags in the main
task string and adds a field in the date bar, which is similar in style
to the due date or the threshold date.

### CODE REVIEW

**Comment by smichel17** [comment]
Is there room for all 3 on a smaller device?

**Comment by TiiXel** [comment]
Here is a screenshot on a 320x480px device. When the date bar is too long, a second line opens, that might be undesired ; maybe having the recurrence pattern alone on a new line, leaving all TextViews unwrapped would be nicer?

screenshot

**Comment by mpcjanssen** [comment]
@smichel17 demonstrated the reservation I have with this pull request. It also breaks the alignment of the other date boxes which I personally find a bit ugly. I am not sure how to best fix it though.

**Comment by smichel17** [comment]
> maybe having the recurrence pattern alone on a new line, leaving all TextViews unwrapped would be nicer?

The downside to that is you're always using extra vertical space, which is valuable on small devices. Really what I'd like is:

- If none of the fields would wrap, leave it as is.
- If any of the fields would wrap, move it to a new line.

I think the wrapping issue is present on the current with larger fonts, so this would improve the current state of things as well. However, I'm not sure if it is possible.

**Comment by mpcjanssen** [comment]
@TiiXel no need to apologise. The whole point of making PRs is so they can be discussed and refined. It's much appreciated.

File: app/src/main/java/nl/mpcjanssen/simpletask/task/TaskAdapter.kt [hunk 1]
```diff
@@ -262,5 +308,26 @@ class TaskAdapter(val completeAction: (Task) -> Unit,
...
+            bottom: Int,
+            oldLeft: Int,
+            oldTop: Int,
+            oldRight: Int,
+            oldBottom: Int
+        ) {
+            v?.removeOnLayoutChangeListener(this)
+            if ((v as TextView).lineCount > 1) {
+                if (!wrapRecurrence) wrapRecurrence = true
+                else if (!wrapThreshold) wrapThreshold = true
```
Comment (smichel17) [inline app/src/main/java/nl/mpcjanssen/simpletask/task/TaskAdapter.kt]: You can get rid of the ifs here and just set both to true.

**Comment by smichel17** [comment]
I'm not sure if there's a way to tell a view to redraw itself, which would likely be the right solution, but from the outside the way to redraw an item is to use the recyclerView's `notifyDataSetChanged()` function (which will update the whole thing, but there's a way to call it so it will only update that one. It's late and I'm on my phone, else I'd link the relevant docs.

**Comment by TiiXel** [comment]
I implemented this if/else because the `else` makes the views wrap one after the other.

If there is room to fit due and threshold dates on the same line, only the recurrence pattern is chopped down.

When the `lineCount > 1` is true again, only then the fields will use three lines.

I believe it worked well, but I can't test it just now.

------------

I tried calling `notifyDataSetChanged()` in the kotlin setter like :

    val wrapRecurrence: Boolean = false
        set(value) {
            notifyDataSetChanged()
            field = value
        }

But it's raising an exception because one can't ask the recycler to redraw when it's already drawing.

**Comment by smichel17** [comment]
> I implemented this if/else because the else makes the views wrap one after the other.

Ha, that's what I get for looking at code in bed at 1:30am :P

I might suggest adding braces and newlines though, so tired-mistakes like that are less likely (scope is clearer).

> Make the list wrap datebars items **on click**

Is this a typo?

> I tried calling `notifyDataSetChanged()` in the kotlin setter [...] but it's raising an exception because one can't ask the recycler to redraw when it's already drawing.

@TiiXel I spent some time looking in to this, and I have some insights, but there's still enough spots I don't understand that it's hard for me to write out a brief and good explanation here. Maybe ping me in irc/matrix/gitter and we can discuss in real time?

**Comment by TiiXel** [comment]
>> Make the list wrap datebars items on click

> Is this a typo?

It's the current undesirable behaviour, I guessed the commit message had to explain what's happening and not what's the objective.

> Maybe ping me in irc/matrix/gitter and we can discuss in real time?

I'll be super busy for a month at least, if I find some time to work on this project I will, sure!

**Comment by mpcjanssen** [comment]
A possible option would be to add an recurring indicator which only shows if the task is recurring but not when. But that will fit better in a complete redo of the task list view.

**Comment by Bellamonte** [comment]
+1

**Comment by mpcjanssen** [comment]
Closing PR, will keep it in mind when redo-ing the display.

**Pull request closed without merging**